src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,350 @@
+  map_zero' := rfl,
+  map_add' := λ a b, by apply quot.sound; apply rel.add_scalar,
+  commutes' := λ r x,
+  begin
+    dsimp only [],
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    apply quot.sound,
+    apply rel.central_scalar,
+  end,
+  smul_def' := λ r x, by refl }
+
+/--
+The universal linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def univ : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, by apply quot.sound; apply rel.add_lin,
+  map_smul' := λ r x, by apply quot.sound; apply rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
+-/
+def lift {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
+    { change f _ = f _ + f _, simp, },
+    { change f _ = (algebra_map _ _ _) * f _, rw linear_map.map_smul,
+      exact algebra.smul_def h_r (f h_a) },
+    { change algebra_map _ _ _ = algebra_map _ _ _ + algebra_map _ _ _,
+      exact (algebra_map R A).map_add h_r h_s },
+    { change algebra_map _ _ _ = algebra_map _ _ _ * algebra_map _ _ _,
+      exact (algebra_map R A).map_mul h_r h_s },
+    { let G := lift_fun R M f,
+      change (algebra_map _ _ _) * G _ = G _,
+      exact algebra.commutes h_r (G h_a) },
+    { change _ + _ + _ = _ + (_ + _), rw add_assoc },
+    { change _ + _ = _ + _, rw add_comm, },
+    { change (algebra_map _ _ _) * _ + _ = algebra_map _ _ _, simp },
+    { let G := lift_fun R M f,
+      change (algebra_map _ _ _) + G _ = G _, simp, },
+    { change _ * _ * _ = _ * (_ * _), rw mul_assoc },
+    { let G := lift_fun R M f,
+      change (algebra_map _ _ _) * G _ = G _, simp, },
+    { let G := lift_fun R M f,
+      change G _ * (algebra_map _ _ _)= G _, simp, },
+    { change _ * (_ + _) = _ * _ + _ * _, rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _, rw right_distrib, },
+    repeat { set G := lift_fun R M f,
+      change G _ + G _ = G _ + G _, rw h_ih, },
+    repeat { set G := lift_fun R M f,
+      change G _ * G _ = G _ * G _, rw h_ih, },
+  end,
+  map_one' := by change algebra_map _ _ _ = _; simp,
+  map_mul' :=
+  begin
+    intros x y,
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    rcases quot.exists_rep y with ⟨y,rfl⟩,
+    refl,
+  end,
+  map_zero' := by change algebra_map _ _ _ = _; simp,
+  map_add' :=
+  begin
+    intros x y,
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    rcases quot.exists_rep y with ⟨y,rfl⟩,
+    refl,
+  end,
+  commutes' := by tauto }
+
+variables {R M}
+
+theorem univ_comp_lift {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) :
+  (lift R M f) ∘ (univ R M) = f := rfl
+
+theorem lift_unique {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A)
+  (g : tensor_algebra R M →ₐ[R] A) : g ∘ (univ R M) = f → g = lift R M f :=
```

Comment:
It might be nice to have a extensionality theorem for maps out of the `tensor_algebra` as well as this. I would state the theorem something like
```lean
theorem hom_ext {A : Type u3} [ring A] [algebra R A] 
  {f g : tensor_algebra R M →ₐ[R] A} : 
  f.to_linear_map.comp (univ R M) = g.to_linear_map.comp (univ R M) → 
  f = g
```
This isn't standard mathlib style yet, but it's something I think is quite useful.

Reply:
I like this. Adding it!